Skip protoc execution when tests are skipped

We tried to use Jitpack to test a build of the Zipkin Server from a pull request (#2328), but the build failed with the following.

https://jitpack.io/com/github/ewhauser/zipkin/experimental-grpc-1.5.0-g7a0c3ce-1160/build.log

```
[INFO] --- protobuf-maven-plugin:0.6.1:test-compile (test-compile-proto) @ zipkin-server ---
[INFO] Compiling 1 proto file(s) to /home/jitpack/build/zipkin-server/target/generated-test-sources/protobuf/java
[ERROR] PROTOC FAILED: /home/jitpack/build/zipkin-server/target/protoc-plugins/protoc-3.7.0-linux-x86_64.exe: error while loading shared libraries: libatomic.so.1: cannot open shared object file: No such file or directory

[ERROR] /home/jitpack/build/zipkin-server/target/test/proto/zipkin.proto [0:0]: /home/jitpack/build/zipkin-server/target/protoc-plugins/protoc-3.7.0-linux-x86_64.exe: error while loading shared libraries: libatomic.so.1: cannot open shared object file: No such file or directory
```

Since this protoc execution is only for tests and the build is skipping tests, we should be able to disable it when tests are skipped so Jitpack succeeds. This can probably be done in a Maven profile.
Alternatively, we can customize the command Jitpack uses to build, but I do like it working out-of-the-box (as it did before the pull request added GRPC collector support) and I wonder if Zipkin developers wouldn't also run into this problem on their machines.
/cc @bsideup

--- a/benchmarks/pom.xml
+++ b/benchmarks/pom.xml
@@ -45,6 +45,11 @@
       <artifactId>guava</artifactId>
       <version>${guava.version}</version>
     </dependency>
+    <!-- for codec benchmarks -->
+    <dependency>
+      <groupId>com.google.code.gson</groupId>
+      <artifactId>gson</artifactId>
+    </dependency>
 
     <dependency>
       <groupId>com.esotericsoftware.kryo</groupId>

--- a/benchmarks/pom.xml
+++ b/benchmarks/pom.xml
@@ -69,19 +69,26 @@
       <artifactId>zipkin-server</artifactId>
       <version>${project.version}</version>
     </dependency>
-
+    <dependency>
+      <groupId>com.squareup.wire</groupId>
+      <artifactId>wire-runtime</artifactId>
+      <version>${wire.version}</version>
+    </dependency>
     <dependency>
       <groupId>io.zipkin.proto3</groupId>
       <artifactId>zipkin-proto3</artifactId>
-    </dependency>
-    <dependency>
-      <groupId>com.google.protobuf</groupId>
-      <artifactId>protobuf-java</artifactId>
     </dependency>
   </dependencies>
 
   <build>
     <plugins>
+      <plugin>
+        <artifactId>maven-dependency-plugin</artifactId>
+      </plugin>
+      <plugin>
+        <groupId>com.squareup.wire</groupId>
+        <artifactId>wire-maven-plugin</artifactId>
+      </plugin>
       <!-- disable retrolambda as we allow language level 1.8 benchmark classes -->
       <plugin>
         <groupId>net.orfjackal.retrolambda</groupId>
@@ -104,27 +111,6 @@
         <configuration>
           <skip>true</skip>
         </configuration>
-      </plugin>
-      <!-- protobuf-maven-plugin cannot get proto definitions from dependencies, so we will -->
-      <plugin>
-        <artifactId>maven-dependency-plugin</artifactId>
-        <executions>
-          <execution>
-            <id>unpack-proto</id>
-          </execution>
-        </executions>
-      </plugin>
-      <plugin>
-        <groupId>org.xolstice.maven.plugins</groupId>
-        <artifactId>protobuf-maven-plugin</artifactId>
-        <executions>
-          <execution>
-            <id>compile-proto</id>
-            <goals>
-              <goal>compile</goal>
-            </goals>
-          </execution>
-        </executions>
       </plugin>
       <plugin>
         <artifactId>maven-shade-plugin</artifactId>

--- a/benchmarks/src/main/java/zipkin2/codec/CodecBenchmarks.java
+++ b/benchmarks/src/main/java/zipkin2/codec/CodecBenchmarks.java
@@ -74,12 +74,12 @@ public class CodecBenchmarks {
 
   /** manually implemented with json so not as slow as normal java */
   @Benchmark
-  public Span readClientSpan_java() {
+  public Span readClientSpan_kryo() {
     return kryo.readObject(new Input(zipkin2Serialized), Span.class);
   }
 
   @Benchmark
-  public byte[] writeClientSpan_java() {
+  public byte[] writeClientSpan_kryo() {
     Output output = new Output(zipkin2Serialized.length);
     kryo.writeObject(output, zipkin2);
     output.flush();
@@ -117,12 +117,12 @@ public class CodecBenchmarks {
   }
 
   @Benchmark
-  public byte[] writeClientSpan_json_legacy() {
+  public byte[] writeClientSpan_json_v1() {
     return SpanBytesEncoder.JSON_V1.encode(zipkin2);
   }
 
   @Benchmark
-  public byte[] writeTenClientSpans_json_legacy() {
+  public byte[] writeTenClientSpans_json_v1() {
     return SpanBytesEncoder.JSON_V1.encodeList(tenSpan2s);
   }
 
@@ -146,7 +146,7 @@ public class CodecBenchmarks {
   }
 
   @Benchmark
-  public zipkin2.proto3.Span readChineseSpan_proto3_protobuf() throws Exception {
+  public zipkin2.proto3.Span readChineseSpan_proto3_wire() throws Exception {
     return ADAPTER.decode(zipkin2Proto3Chinese);
   }
 

--- a/benchmarks/src/main/java/zipkin2/codec/CodecBenchmarks.java
+++ b/benchmarks/src/main/java/zipkin2/codec/CodecBenchmarks.java
@@ -41,7 +41,7 @@ import org.openjdk.jmh.runner.options.Options;
 import org.openjdk.jmh.runner.options.OptionsBuilder;
 import zipkin2.Span;
 
-import static zipkin2.proto3.Span.parseFrom;
+import static zipkin2.proto3.Span.ADAPTER;
 
 /**
  * The {@link SpanBytesEncoder bundled java codec} aims to be both small in size (i.e. does not
@@ -97,8 +97,8 @@ public class CodecBenchmarks {
   }
 
   @Benchmark
-  public zipkin2.proto3.Span readClientSpan_proto3_protobuf() throws Exception {
-    return parseFrom(zipkin2Proto3);
+  public zipkin2.proto3.Span readClientSpan_proto3_wire() throws Exception {
+    return ADAPTER.decode(zipkin2Proto3);
   }
 
   @Benchmark
@@ -147,7 +147,7 @@ public class CodecBenchmarks {
 
   @Benchmark
   public zipkin2.proto3.Span readChineseSpan_proto3_protobuf() throws Exception {
-    return parseFrom(zipkin2Proto3Chinese);
+    return ADAPTER.decode(zipkin2Proto3Chinese);
   }
 
   @Benchmark

--- a/benchmarks/src/main/java/zipkin2/codec/CodecBenchmarks.java
+++ b/benchmarks/src/main/java/zipkin2/codec/CodecBenchmarks.java
@@ -41,12 +41,15 @@ import org.openjdk.jmh.runner.options.Options;
 import org.openjdk.jmh.runner.options.OptionsBuilder;
 import zipkin2.Span;
 
-import static zipkin2.proto3.Span.ADAPTER;
-
 /**
  * The {@link SpanBytesEncoder bundled java codec} aims to be both small in size (i.e. does not
  * significantly increase the size of zipkin's jar), and efficient. It may not always be fastest,
  * but we should try to keep it competitive.
+ *
+ * <p>Note that the wire benchmarks use their structs, not ours. This will result in more efficient
+ * writes as there's no hex codec of IDs, stringifying of IPs etc. A later change could do that, but
+ * it likely still going to be more efficient than our dependency-free codec. This means in cases
+ * where extra dependencies are ok (such as our server), we could consider using wire.
  */
 @Measurement(iterations = 5, time = 1)
 @Warmup(iterations = 10, time = 1)
@@ -56,108 +59,133 @@ import static zipkin2.proto3.Span.ADAPTER;
 @State(Scope.Thread)
 @Threads(1)
 public class CodecBenchmarks {
-  static final byte[] zipkin2Json = read("/zipkin2-client.json");
-  static final Span zipkin2 = SpanBytesDecoder.JSON_V2.decodeOne(zipkin2Json);
-  static final byte[] zipkin2Proto3 = SpanBytesEncoder.PROTO3.encode(zipkin2);
-  static final List<Span> tenSpan2s = Collections.nCopies(10, zipkin2);
-  static final byte[] tenSpan2sJson = SpanBytesEncoder.JSON_V2.encodeList(tenSpan2s);
+  static final byte[] clientSpanJsonV2 = read("/zipkin2-client.json");
+  static final Span clientSpan = SpanBytesDecoder.JSON_V2.decodeOne(clientSpanJsonV2);
+  static final byte[] clientSpanProto3 = SpanBytesEncoder.PROTO3.encode(clientSpan);
+  static final zipkin2.proto3.Span clientSpan_wire;
+  static final List<Span> tenClientSpans = Collections.nCopies(10, clientSpan);
+  static final byte[] tenClientSpansJsonV2 = SpanBytesEncoder.JSON_V2.encodeList(tenClientSpans);
   static final Kryo kryo = new Kryo();
-  static final byte[] zipkin2Serialized;
+  static final byte[] clientSpanSerialized;
 
   static {
     kryo.register(Span.class, new JavaSerializer());
     Output output = new Output(4096);
-    kryo.writeObject(output, zipkin2);
+    kryo.writeObject(output, clientSpan);
     output.flush();
-    zipkin2Serialized = output.getBuffer();
+    clientSpanSerialized = output.getBuffer();
+    try {
+      clientSpan_wire = zipkin2.proto3.Span.ADAPTER.decode(clientSpanProto3);
+    } catch (IOException e) {
+      throw new AssertionError(e);
+    }
   }
 
   /** manually implemented with json so not as slow as normal java */
   @Benchmark
   public Span readClientSpan_kryo() {
-    return kryo.readObject(new Input(zipkin2Serialized), Span.class);
+    return kryo.readObject(new Input(clientSpanSerialized), Span.class);
   }
 
   @Benchmark
   public byte[] writeClientSpan_kryo() {
-    Output output = new Output(zipkin2Serialized.length);
-    kryo.writeObject(output, zipkin2);
+    Output output = new Output(clientSpanSerialized.length);
+    kryo.writeObject(output, clientSpan);
     output.flush();
     return output.getBuffer();
   }
 
   @Benchmark
   public Span readClientSpan_json() {
-    return SpanBytesDecoder.JSON_V2.decodeOne(zipkin2Json);
+    return SpanBytesDecoder.JSON_V2.decodeOne(clientSpanJsonV2);
   }
 
   @Benchmark
   public Span readClientSpan_proto3() {
-    return SpanBytesDecoder.PROTO3.decodeOne(zipkin2Proto3);
+    return SpanBytesDecoder.PROTO3.decodeOne(clientSpanProto3);
   }
 
   @Benchmark
   public zipkin2.proto3.Span readClientSpan_proto3_wire() throws Exception {
-    return ADAPTER.decode(zipkin2Proto3);
+    return zipkin2.proto3.Span.ADAPTER.decode(clientSpanProto3);
   }
 
   @Benchmark
   public List<Span> readTenClientSpans_json() {
-    return SpanBytesDecoder.JSON_V2.decodeList(tenSpan2sJson);
+    return SpanBytesDecoder.JSON_V2.decodeList(tenClientSpansJsonV2);
   }
 
   @Benchmark
   public byte[] writeClientSpan_json() {
-    return SpanBytesEncoder.JSON_V2.encode(zipkin2);
+    return SpanBytesEncoder.JSON_V2.encode(clientSpan);
   }
 
   @Benchmark
   public byte[] writeTenClientSpans_json() {
-    return SpanBytesEncoder.JSON_V2.encodeList(tenSpan2s);
+    return SpanBytesEncoder.JSON_V2.encodeList(tenClientSpans);
   }
 
   @Benchmark
   public byte[] writeClientSpan_json_v1() {
-    return SpanBytesEncoder.JSON_V1.encode(zipkin2);
+    return SpanBytesEncoder.JSON_V1.encode(clientSpan);
   }
 
   @Benchmark
   public byte[] writeTenClientSpans_json_v1() {
-    return SpanBytesEncoder.JSON_V1.encodeList(tenSpan2s);
+    return SpanBytesEncoder.JSON_V1.encodeList(tenClientSpans);
   }
 
   @Benchmark
   public byte[] writeClientSpan_proto3() {
-    return SpanBytesEncoder.PROTO3.encode(zipkin2);
+    return SpanBytesEncoder.PROTO3.encode(clientSpan);
   }
 
-  static final byte[] zipkin2JsonChinese = read("/zipkin2-chinese.json");
-  static final Span zipkin2Chinese = SpanBytesDecoder.JSON_V2.decodeOne(zipkin2JsonChinese);
-  static final byte[] zipkin2Proto3Chinese = SpanBytesEncoder.PROTO3.encode(zipkin2Chinese);
+  @Benchmark
+  public byte[] writeClientSpan_proto3_wire() {
+    return clientSpan_wire.encode();
+  }
+
+  static final byte[] chineseSpanJsonV2 = read("/zipkin2-chinese.json");
+  static final Span chineseSpan = SpanBytesDecoder.JSON_V2.decodeOne(chineseSpanJsonV2);
+  static final zipkin2.proto3.Span chineseSpan_wire;
+  static final byte[] chineseSpanProto3 = SpanBytesEncoder.PROTO3.encode(chineseSpan);
+
+  static {
+    try {
+      chineseSpan_wire = zipkin2.proto3.Span.ADAPTER.decode(chineseSpanProto3);
+    } catch (IOException e) {
+      throw new AssertionError(e);
+    }
+  }
 
   @Benchmark
   public Span readChineseSpan_json() {
-    return SpanBytesDecoder.JSON_V2.decodeOne(zipkin2JsonChinese);
+    return SpanBytesDecoder.JSON_V2.decodeOne(chineseSpanJsonV2);
   }
 
   @Benchmark
   public Span readChineseSpan_proto3() {
-    return SpanBytesDecoder.PROTO3.decodeOne(zipkin2Proto3Chinese);
+    return SpanBytesDecoder.PROTO3.decodeOne(chineseSpanProto3);
   }
 
   @Benchmark
   public zipkin2.proto3.Span readChineseSpan_proto3_wire() throws Exception {
-    return ADAPTER.decode(zipkin2Proto3Chinese);
+    return zipkin2.proto3.Span.ADAPTER.decode(chineseSpanProto3);
   }
 
   @Benchmark
   public byte[] writeChineseSpan_json() {
-    return SpanBytesEncoder.JSON_V2.encode(zipkin2Chinese);
+    return SpanBytesEncoder.JSON_V2.encode(chineseSpan);
   }
 
   @Benchmark
   public byte[] writeChineseSpan_proto3() {
-    return SpanBytesEncoder.PROTO3.encode(zipkin2Chinese);
+    return SpanBytesEncoder.PROTO3.encode(chineseSpan);
+  }
+
+  @Benchmark
+  public byte[] writeChineseSpan_proto3_wire() {
+    return chineseSpan_wire.encode();
   }
 
   // Convenience main entry-point

--- a/benchmarks/src/test/java/zipkin2/internal/Proto3CodecInteropTest.java
+++ b/benchmarks/src/test/java/zipkin2/internal/Proto3CodecInteropTest.java
@@ -16,25 +16,23 @@
  */
 package zipkin2.internal;
 
-import com.google.common.io.BaseEncoding;
-import com.google.protobuf.ByteString;
-import com.google.protobuf.CodedOutputStream;
+import com.squareup.wire.ProtoWriter;
 import java.io.IOException;
-import java.util.Arrays;
 import java.util.List;
+import okio.ByteString;
 import org.assertj.core.data.MapEntry;
 import org.junit.Test;
 import zipkin2.codec.SpanBytesDecoder;
 import zipkin2.codec.SpanBytesEncoder;
-import zipkin2.internal.Proto3Fields.Utf8Field;
-import zipkin2.internal.Proto3ZipkinFields.EndpointField;
 import zipkin2.internal.Proto3ZipkinFields.TagField;
 import zipkin2.proto3.Annotation;
 import zipkin2.proto3.Endpoint;
 import zipkin2.proto3.ListOfSpans;
 import zipkin2.proto3.Span;
 
-import static com.google.protobuf.CodedOutputStream.computeMessageSize;
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonMap;
+import static okio.ByteString.decodeHex;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.data.MapEntry.entry;
 import static zipkin2.internal.Proto3ZipkinFields.SPAN;
@@ -43,6 +41,7 @@ import static zipkin2.internal.Proto3ZipkinFields.SpanField.LOCAL_ENDPOINT;
 import static zipkin2.internal.Proto3ZipkinFields.SpanField.REMOTE_ENDPOINT;
 import static zipkin2.internal.Proto3ZipkinFields.SpanField.TAG_KEY;
 
+// Compares against Square Wire as it is easier than Google's protobuf tooling
 public class Proto3CodecInteropTest {
   static final zipkin2.Endpoint ORDER = zipkin2.Endpoint.newBuilder()
     .serviceName("订单维护服务")
@@ -72,108 +71,104 @@ public class Proto3CodecInteropTest {
     .putTag("error", "此用户没有操作权限")
     .shared(true)
     .build();
-  static final List<zipkin2.Span> ZIPKIN_SPANS = Arrays.asList(ZIPKIN_SPAN, ZIPKIN_SPAN);
+  static final List<zipkin2.Span> ZIPKIN_SPANS = asList(ZIPKIN_SPAN, ZIPKIN_SPAN);
 
-  static final Span PROTO_SPAN = Span.newBuilder()
-    .setTraceId(decodeHex(ZIPKIN_SPAN.traceId()))
-    .setParentId(decodeHex(ZIPKIN_SPAN.parentId()))
-    .setId(decodeHex(ZIPKIN_SPAN.id()))
-    .setKind(Span.Kind.valueOf(ZIPKIN_SPAN.kind().name()))
-    .setName(ZIPKIN_SPAN.name())
-    .setTimestamp(ZIPKIN_SPAN.timestampAsLong())
-    .setDuration(ZIPKIN_SPAN.durationAsLong())
-    .setLocalEndpoint(Endpoint.newBuilder()
-      .setServiceName(ORDER.serviceName())
-      .setIpv6(ByteString.copyFrom(ORDER.ipv6Bytes())).build()
+  static final Span PROTO_SPAN = new Span.Builder()
+    .trace_id(decodeHex(ZIPKIN_SPAN.traceId()))
+    .parent_id(decodeHex(ZIPKIN_SPAN.parentId()))
+    .id(decodeHex(ZIPKIN_SPAN.id()))
+    .kind(Span.Kind.valueOf(ZIPKIN_SPAN.kind().name()))
+    .name(ZIPKIN_SPAN.name())
+    .timestamp(ZIPKIN_SPAN.timestampAsLong())
+    .duration(ZIPKIN_SPAN.durationAsLong())
+    .local_endpoint(new Endpoint.Builder()
+      .service_name(ORDER.serviceName())
+      .ipv6(ByteString.of(ORDER.ipv6Bytes())).build()
     )
-    .setRemoteEndpoint(Endpoint.newBuilder()
-      .setServiceName(PROFILE.serviceName())
-      .setIpv4(ByteString.copyFrom(PROFILE.ipv4Bytes()))
-      .setPort(PROFILE.portAsInt()).build()
+    .remote_endpoint(new Endpoint.Builder()
+      .service_name(PROFILE.serviceName())
+      .ipv4(ByteString.of(PROFILE.ipv4Bytes()))
+      .port(PROFILE.portAsInt()).build()
     )
-    .addAnnotations(Annotation.newBuilder()
-      .setTimestamp(ZIPKIN_SPAN.annotations().get(0).timestamp())
-      .setValue(ZIPKIN_SPAN.annotations().get(0).value())
-      .build())
-    .putAllTags(ZIPKIN_SPAN.tags())
-    .setShared(true)
+    .annotations(asList(new Annotation.Builder()
+      .timestamp(ZIPKIN_SPAN.annotations().get(0).timestamp())
+      .value(ZIPKIN_SPAN.annotations().get(0).value())
+      .build()))
+    .tags(ZIPKIN_SPAN.tags())
+    .shared(true)
     .build();
-  ListOfSpans PROTO_SPANS = ListOfSpans.newBuilder()
-    .addSpans(PROTO_SPAN)
-    .addSpans(PROTO_SPAN).build();
+  ListOfSpans PROTO_SPANS = new ListOfSpans.Builder()
+    .spans(asList(PROTO_SPAN, PROTO_SPAN)).build();
 
-  @Test public void encodeIsCompatible() throws Exception {
-    byte[] googleBytes = new byte[computeMessageSize(1, PROTO_SPAN)];
-    CodedOutputStream out = CodedOutputStream.newInstance(googleBytes);
-    out.writeMessage(1, PROTO_SPAN);
+  @Test public void encodeIsCompatible() throws IOException {
+    okio.Buffer buffer = new okio.Buffer();
+
+    Span.ADAPTER.encodeWithTag(new ProtoWriter(buffer), 1, PROTO_SPAN);
 
     assertThat(SpanBytesEncoder.PROTO3.encode(ZIPKIN_SPAN))
-      .containsExactly(googleBytes);
+      .containsExactly(buffer.readByteArray());
   }
 
   @Test public void decodeOneIsCompatible() {
-    assertThat(SpanBytesDecoder.PROTO3.decodeOne(PROTO_SPANS.toByteArray()))
+    assertThat(SpanBytesDecoder.PROTO3.decodeOne(PROTO_SPANS.encode()))
       .isEqualTo(ZIPKIN_SPAN);
   }
 
   @Test public void decodeListIsCompatible() {
-    assertThat(SpanBytesDecoder.PROTO3.decodeList(PROTO_SPANS.toByteArray()))
+    assertThat(SpanBytesDecoder.PROTO3.decodeList(PROTO_SPANS.encode()))
       .containsExactly(ZIPKIN_SPAN, ZIPKIN_SPAN);
   }
 
-  @Test public void encodeListIsCompatible_buff() throws Exception {
-    byte[] googleBytes = new byte[PROTO_SPANS.getSerializedSize()];
-    CodedOutputStream out = CodedOutputStream.newInstance(googleBytes);
-    PROTO_SPANS.writeTo(out);
+  @Test public void encodeListIsCompatible_buff() {
+    byte[] wireBytes = PROTO_SPANS.encode();
+    byte[] zipkin_buff = new byte[10 + wireBytes.length];
 
-    byte[] zipkin_buff = new byte[10 + googleBytes.length];
     assertThat(SpanBytesEncoder.PROTO3.encodeList(ZIPKIN_SPANS, zipkin_buff, 5))
-      .isEqualTo(googleBytes.length);
+      .isEqualTo(wireBytes.length);
 
     assertThat(zipkin_buff)
       .startsWith(0, 0, 0, 0, 0)
-      .containsSequence(googleBytes)
+      .containsSequence(wireBytes)
       .endsWith(0, 0, 0, 0, 0);
   }
 
-  @Test public void encodeListIsCompatible() throws Exception {
-    byte[] googleBytes = new byte[PROTO_SPANS.getSerializedSize()];
-    CodedOutputStream out = CodedOutputStream.newInstance(googleBytes);
-    PROTO_SPANS.writeTo(out);
+  @Test public void encodeListIsCompatible() {
+    byte[] wireBytes = PROTO_SPANS.encode();
 
     assertThat(SpanBytesEncoder.PROTO3.encodeList(ZIPKIN_SPANS))
-      .containsExactly(googleBytes);
+      .containsExactly(wireBytes);
   }
 
-  @Test public void span_sizeInBytes_matchesProto3() {
+  @Test public void span_sizeInBytes_matchesWire() {
     assertThat(SPAN.sizeInBytes(ZIPKIN_SPAN))
-      .isEqualTo(computeMessageSize(SPAN.fieldNumber, PROTO_SPAN));
+      .isEqualTo(Span.ADAPTER.encodedSizeWithTag(SPAN.fieldNumber, PROTO_SPAN));
   }
 
-  @Test public void annotation_sizeInBytes_matchesProto3() {
+  @Test public void annotation_sizeInBytes_matchesWire() {
     zipkin2.Annotation zipkinAnnotation = ZIPKIN_SPAN.annotations().get(0);
 
     assertThat(ANNOTATION.sizeInBytes(zipkinAnnotation))
-      .isEqualTo(computeMessageSize(ANNOTATION.fieldNumber, PROTO_SPAN.getAnnotations(0)));
+      .isEqualTo(Annotation.ADAPTER.encodedSizeWithTag(ANNOTATION.fieldNumber,
+        PROTO_SPAN.annotations.get(0)));
   }
 
-  @Test public void annotation_write_matchesProto3() throws IOException {
+  @Test public void annotation_write_matchesWire() {
     zipkin2.Annotation zipkinAnnotation = ZIPKIN_SPAN.annotations().get(0);
-    Annotation protoAnnotation = PROTO_SPAN.getAnnotations(0);
+    Annotation protoAnnotation = PROTO_SPAN.annotations.get(0);
 
     Buffer zipkinBytes = Buffer.allocate(ANNOTATION.sizeInBytes(zipkinAnnotation));
     ANNOTATION.write(zipkinBytes, zipkinAnnotation);
 
     assertThat(zipkinBytes.toByteArray())
-      .containsExactly(writeSpan(Span.newBuilder().addAnnotations(protoAnnotation).build()));
+      .containsExactly(new Span.Builder().annotations(asList(protoAnnotation)).build().encode());
   }
 
-  @Test public void annotation_read_matchesProto3() throws IOException {
+  @Test public void annotation_read_matchesWire() {
     zipkin2.Annotation zipkinAnnotation = ZIPKIN_SPAN.annotations().get(0);
-    Annotation protoAnnotation = PROTO_SPAN.getAnnotations(0);
+    Annotation protoAnnotation = PROTO_SPAN.annotations.get(0);
 
     Buffer zipkinBytes =
-      Buffer.wrap(writeSpan(Span.newBuilder().addAnnotations(protoAnnotation).build()), 0);
+      Buffer.wrap(new Span.Builder().annotations(asList(protoAnnotation)).build().encode(), 0);
     assertThat(zipkinBytes.readVarint32())
       .isEqualTo(ANNOTATION.key);
 
@@ -183,81 +178,62 @@ public class Proto3CodecInteropTest {
       .containsExactly(zipkinAnnotation);
   }
 
-  @Test public void endpoint_sizeInBytes_matchesProto3() {
+  @Test public void endpoint_sizeInBytes_matchesWire() {
     assertThat(LOCAL_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.localEndpoint()))
-      .isEqualTo(computeMessageSize(LOCAL_ENDPOINT.fieldNumber, PROTO_SPAN.getLocalEndpoint()));
+      .isEqualTo(
+        Endpoint.ADAPTER.encodedSizeWithTag(LOCAL_ENDPOINT.fieldNumber, PROTO_SPAN.local_endpoint));
 
     assertThat(REMOTE_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.remoteEndpoint()))
-      .isEqualTo(computeMessageSize(REMOTE_ENDPOINT.fieldNumber, PROTO_SPAN.getRemoteEndpoint()));
+      .isEqualTo(Endpoint.ADAPTER.encodedSizeWithTag(REMOTE_ENDPOINT.fieldNumber,
+        PROTO_SPAN.remote_endpoint));
   }
 
-  @Test public void localEndpoint_write_matchesProto3() throws IOException {
+  @Test public void localEndpoint_write_matchesWire() {
     Buffer zipkinBytes = Buffer.allocate(LOCAL_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.localEndpoint()));
     LOCAL_ENDPOINT.write(zipkinBytes, ZIPKIN_SPAN.localEndpoint());
+    Span span = new Span.Builder().local_endpoint(PROTO_SPAN.local_endpoint).build();
 
     assertThat(zipkinBytes.toByteArray())
-      .containsExactly(
-        writeSpan(Span.newBuilder().setLocalEndpoint(PROTO_SPAN.getLocalEndpoint()).build()));
+      .containsExactly(span.encode());
   }
 
-  @Test public void remoteEndpoint_write_matchesProto3() throws IOException {
+  @Test public void remoteEndpoint_write_matchesWire() {
     Buffer zipkinBytes = Buffer.allocate(REMOTE_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.remoteEndpoint()));
     REMOTE_ENDPOINT.write(zipkinBytes, ZIPKIN_SPAN.remoteEndpoint());
+    Span span = new Span.Builder().remote_endpoint(PROTO_SPAN.remote_endpoint).build();
 
     assertThat(zipkinBytes.toByteArray())
-      .containsExactly(
-        writeSpan(Span.newBuilder().setRemoteEndpoint(PROTO_SPAN.getRemoteEndpoint()).build()));
+      .containsExactly(span.encode());
   }
 
-  @Test public void utf8_sizeInBytes_matchesProto3() {
-    assertThat(new Utf8Field(EndpointField.SERVICE_NAME_KEY).sizeInBytes(ORDER.serviceName()))
-      .isEqualTo(CodedOutputStream.computeStringSize(1, ORDER.serviceName()));
-  }
-
-  @Test public void tag_sizeInBytes_matchesProto3() {
+  @Test public void tag_sizeInBytes_matchesWire() {
     MapEntry<String, String> entry = entry("clnt/finagle.version", "6.45.0");
+    Span span = new Span.Builder().tags(singletonMap(entry.key, entry.value)).build();
+
     assertThat(new TagField(TAG_KEY).sizeInBytes(entry))
-      .isEqualTo(Span.newBuilder().putTags(entry.key, entry.value).build().getSerializedSize());
+      .isEqualTo(Span.ADAPTER.encodedSize(span));
   }
 
-  @Test public void writeTagField_matchesProto3() throws IOException {
+  @Test public void writeTagField_matchesWire() {
     MapEntry<String, String> entry = entry("clnt/finagle.version", "6.45.0");
     TagField field = new TagField(TAG_KEY);
     Buffer zipkinBytes = Buffer.allocate(field.sizeInBytes(entry));
     field.write(zipkinBytes, entry);
 
-    Span oneField = Span.newBuilder().putTags(entry.key, entry.value).build();
-    byte[] googleBytes = new byte[oneField.getSerializedSize()];
-    CodedOutputStream out = CodedOutputStream.newInstance(googleBytes);
-    oneField.writeTo(out);
-
+    Span oneField = new Span.Builder().tags(singletonMap(entry.key, entry.value)).build();
     assertThat(zipkinBytes.toByteArray())
-      .containsExactly(googleBytes);
+      .containsExactly(oneField.encode());
   }
 
-  @Test public void writeTagField_matchesProto3_emptyValue() throws IOException {
+  @Test public void writeTagField_matchesWire_emptyValue() {
     MapEntry<String, String> entry = entry("error", "");
     TagField field = new TagField(TAG_KEY);
     Buffer zipkinBytes = Buffer.allocate(field.sizeInBytes(entry));
     field.write(zipkinBytes, entry);
 
-    Span oneField = Span.newBuilder().putTags(entry.key, entry.value).build();
-    byte[] googleBytes = new byte[oneField.getSerializedSize()];
-    CodedOutputStream out = CodedOutputStream.newInstance(googleBytes);
-    oneField.writeTo(out);
-
+    Span oneField = new Span.Builder().tags(singletonMap(entry.key, entry.value)).build();
     assertThat(zipkinBytes.toByteArray())
-      .containsExactly(googleBytes);
-  }
-
-  static ByteString decodeHex(String s) {
-    return ByteString.copyFrom(BaseEncoding.base16().lowerCase().decode(s));
-  }
-
-  static byte[] writeSpan(Span span) throws IOException {
-    byte[] googleBytes = new byte[span.getSerializedSize()];
-    span.writeTo(CodedOutputStream.newInstance(googleBytes));
-    return googleBytes;
+      .containsExactly(oneField.encode());
   }
 
   static zipkin2.Span.Builder zipkinSpanBuilder() {

--- a/benchmarks/src/test/java/zipkin2/internal/Proto3CodecInteropTest.java
+++ b/benchmarks/src/test/java/zipkin2/internal/Proto3CodecInteropTest.java
@@ -18,6 +18,7 @@ package zipkin2.internal;
 
 import com.squareup.wire.ProtoWriter;
 import java.io.IOException;
+import java.util.Arrays;
 import java.util.List;
 import okio.ByteString;
 import org.assertj.core.data.MapEntry;
@@ -30,7 +31,6 @@ import zipkin2.proto3.Endpoint;
 import zipkin2.proto3.ListOfSpans;
 import zipkin2.proto3.Span;
 
-import static java.util.Arrays.asList;
 import static java.util.Collections.singletonMap;
 import static okio.ByteString.decodeHex;
 import static org.assertj.core.api.Assertions.assertThat;
@@ -71,7 +71,7 @@ public class Proto3CodecInteropTest {
     .putTag("error", "此用户没有操作权限")
     .shared(true)
     .build();
-  static final List<zipkin2.Span> ZIPKIN_SPANS = asList(ZIPKIN_SPAN, ZIPKIN_SPAN);
+  static final List<zipkin2.Span> ZIPKIN_SPANS = Arrays.asList(ZIPKIN_SPAN, ZIPKIN_SPAN);
 
   static final Span PROTO_SPAN = new Span.Builder()
     .trace_id(decodeHex(ZIPKIN_SPAN.traceId()))
@@ -90,7 +90,7 @@ public class Proto3CodecInteropTest {
       .ipv4(ByteString.of(PROFILE.ipv4Bytes()))
       .port(PROFILE.portAsInt()).build()
     )
-    .annotations(asList(new Annotation.Builder()
+    .annotations(Arrays.asList(new Annotation.Builder()
       .timestamp(ZIPKIN_SPAN.annotations().get(0).timestamp())
       .value(ZIPKIN_SPAN.annotations().get(0).value())
       .build()))
@@ -98,7 +98,7 @@ public class Proto3CodecInteropTest {
     .shared(true)
     .build();
   ListOfSpans PROTO_SPANS = new ListOfSpans.Builder()
-    .spans(asList(PROTO_SPAN, PROTO_SPAN)).build();
+    .spans(Arrays.asList(PROTO_SPAN, PROTO_SPAN)).build();
 
   @Test public void encodeIsCompatible() throws IOException {
     okio.Buffer buffer = new okio.Buffer();
@@ -141,77 +141,75 @@ public class Proto3CodecInteropTest {
 
   @Test public void span_sizeInBytes_matchesWire() {
     assertThat(SPAN.sizeInBytes(ZIPKIN_SPAN))
-      .isEqualTo(Span.ADAPTER.encodedSizeWithTag(SPAN.fieldNumber, PROTO_SPAN));
+      .isEqualTo(Span.ADAPTER.encodedSizeWithTag(SPAN.key, PROTO_SPAN));
   }
 
   @Test public void annotation_sizeInBytes_matchesWire() {
     zipkin2.Annotation zipkinAnnotation = ZIPKIN_SPAN.annotations().get(0);
 
-    assertThat(ANNOTATION.sizeInBytes(zipkinAnnotation))
-      .isEqualTo(Annotation.ADAPTER.encodedSizeWithTag(ANNOTATION.fieldNumber,
-        PROTO_SPAN.annotations.get(0)));
+    assertThat(ANNOTATION.sizeInBytes(zipkinAnnotation)).isEqualTo(
+      Annotation.ADAPTER.encodedSizeWithTag(ANNOTATION.key, PROTO_SPAN.annotations.get(0)));
   }
 
   @Test public void annotation_write_matchesWire() {
     zipkin2.Annotation zipkinAnnotation = ZIPKIN_SPAN.annotations().get(0);
-    Annotation protoAnnotation = PROTO_SPAN.annotations.get(0);
+    Span wireSpan = new Span.Builder().annotations(PROTO_SPAN.annotations).build();
 
     Buffer zipkinBytes = Buffer.allocate(ANNOTATION.sizeInBytes(zipkinAnnotation));
     ANNOTATION.write(zipkinBytes, zipkinAnnotation);
 
     assertThat(zipkinBytes.toByteArray())
-      .containsExactly(new Span.Builder().annotations(asList(protoAnnotation)).build().encode());
+      .containsExactly(wireSpan.encode());
   }
 
-  @Test public void annotation_read_matchesWire() {
+  @Test public void annotation_read_matchesWireEncodingWithTag() {
     zipkin2.Annotation zipkinAnnotation = ZIPKIN_SPAN.annotations().get(0);
-    Annotation protoAnnotation = PROTO_SPAN.annotations.get(0);
+    Span wireSpan = new Span.Builder().annotations(PROTO_SPAN.annotations).build();
 
-    Buffer zipkinBytes =
-      Buffer.wrap(new Span.Builder().annotations(asList(protoAnnotation)).build().encode(), 0);
-    assertThat(zipkinBytes.readVarint32())
+    Buffer wireBytes = Buffer.wrap(wireSpan.encode(), 0);
+    assertThat(wireBytes.readVarint32())
       .isEqualTo(ANNOTATION.key);
 
     zipkin2.Span.Builder builder = zipkinSpanBuilder();
-    ANNOTATION.readLengthPrefixAndValue(zipkinBytes, builder);
+    ANNOTATION.readLengthPrefixAndValue(wireBytes, builder);
     assertThat(builder.build().annotations())
       .containsExactly(zipkinAnnotation);
   }
 
-  @Test public void endpoint_sizeInBytes_matchesWire() {
-    assertThat(LOCAL_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.localEndpoint()))
-      .isEqualTo(
-        Endpoint.ADAPTER.encodedSizeWithTag(LOCAL_ENDPOINT.fieldNumber, PROTO_SPAN.local_endpoint));
+  @Test public void endpoint_sizeInBytes_matchesWireEncodingWithTag() {
+    assertThat(LOCAL_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.localEndpoint())).isEqualTo(
+        Endpoint.ADAPTER.encodedSizeWithTag(LOCAL_ENDPOINT.key, PROTO_SPAN.local_endpoint)
+    );
 
-    assertThat(REMOTE_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.remoteEndpoint()))
-      .isEqualTo(Endpoint.ADAPTER.encodedSizeWithTag(REMOTE_ENDPOINT.fieldNumber,
-        PROTO_SPAN.remote_endpoint));
+    assertThat(REMOTE_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.remoteEndpoint())).isEqualTo(
+      Endpoint.ADAPTER.encodedSizeWithTag(REMOTE_ENDPOINT.key, PROTO_SPAN.remote_endpoint)
+    );
   }
 
   @Test public void localEndpoint_write_matchesWire() {
     Buffer zipkinBytes = Buffer.allocate(LOCAL_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.localEndpoint()));
     LOCAL_ENDPOINT.write(zipkinBytes, ZIPKIN_SPAN.localEndpoint());
-    Span span = new Span.Builder().local_endpoint(PROTO_SPAN.local_endpoint).build();
+    Span wireSpan = new Span.Builder().local_endpoint(PROTO_SPAN.local_endpoint).build();
 
     assertThat(zipkinBytes.toByteArray())
-      .containsExactly(span.encode());
+      .containsExactly(wireSpan.encode());
   }
 
   @Test public void remoteEndpoint_write_matchesWire() {
     Buffer zipkinBytes = Buffer.allocate(REMOTE_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.remoteEndpoint()));
     REMOTE_ENDPOINT.write(zipkinBytes, ZIPKIN_SPAN.remoteEndpoint());
-    Span span = new Span.Builder().remote_endpoint(PROTO_SPAN.remote_endpoint).build();
+    Span wireSpan = new Span.Builder().remote_endpoint(PROTO_SPAN.remote_endpoint).build();
 
     assertThat(zipkinBytes.toByteArray())
-      .containsExactly(span.encode());
+      .containsExactly(wireSpan.encode());
   }
 
   @Test public void tag_sizeInBytes_matchesWire() {
     MapEntry<String, String> entry = entry("clnt/finagle.version", "6.45.0");
-    Span span = new Span.Builder().tags(singletonMap(entry.key, entry.value)).build();
+    Span wireSpan = new Span.Builder().tags(singletonMap(entry.key, entry.value)).build();
 
     assertThat(new TagField(TAG_KEY).sizeInBytes(entry))
-      .isEqualTo(Span.ADAPTER.encodedSize(span));
+      .isEqualTo(Span.ADAPTER.encodedSize(wireSpan));
   }
 
   @Test public void writeTagField_matchesWire() {

--- a/benchmarks/src/test/java/zipkin2/internal/Proto3CodecInteropTest.java
+++ b/benchmarks/src/test/java/zipkin2/internal/Proto3CodecInteropTest.java
@@ -141,14 +141,15 @@ public class Proto3CodecInteropTest {
 
   @Test public void span_sizeInBytes_matchesWire() {
     assertThat(SPAN.sizeInBytes(ZIPKIN_SPAN))
-      .isEqualTo(Span.ADAPTER.encodedSizeWithTag(SPAN.key, PROTO_SPAN));
+      .isEqualTo(Span.ADAPTER.encodedSizeWithTag(SPAN.fieldNumber, PROTO_SPAN));
   }
 
   @Test public void annotation_sizeInBytes_matchesWire() {
     zipkin2.Annotation zipkinAnnotation = ZIPKIN_SPAN.annotations().get(0);
 
     assertThat(ANNOTATION.sizeInBytes(zipkinAnnotation)).isEqualTo(
-      Annotation.ADAPTER.encodedSizeWithTag(ANNOTATION.key, PROTO_SPAN.annotations.get(0)));
+      Annotation.ADAPTER.encodedSizeWithTag(ANNOTATION.fieldNumber, PROTO_SPAN.annotations.get(0))
+    );
   }
 
   @Test public void annotation_write_matchesWire() {
@@ -178,11 +179,11 @@ public class Proto3CodecInteropTest {
 
   @Test public void endpoint_sizeInBytes_matchesWireEncodingWithTag() {
     assertThat(LOCAL_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.localEndpoint())).isEqualTo(
-        Endpoint.ADAPTER.encodedSizeWithTag(LOCAL_ENDPOINT.key, PROTO_SPAN.local_endpoint)
+        Endpoint.ADAPTER.encodedSizeWithTag(LOCAL_ENDPOINT.fieldNumber, PROTO_SPAN.local_endpoint)
     );
 
     assertThat(REMOTE_ENDPOINT.sizeInBytes(ZIPKIN_SPAN.remoteEndpoint())).isEqualTo(
-      Endpoint.ADAPTER.encodedSizeWithTag(REMOTE_ENDPOINT.key, PROTO_SPAN.remote_endpoint)
+      Endpoint.ADAPTER.encodedSizeWithTag(REMOTE_ENDPOINT.fieldNumber, PROTO_SPAN.remote_endpoint)
     );
   }
 

--- a/pom.xml
+++ b/pom.xml
@@ -389,13 +389,6 @@
   </dependencies>
 
   <build>
-    <extensions>
-      <extension>
-        <groupId>kr.motd.maven</groupId>
-        <artifactId>os-maven-plugin</artifactId>
-        <version>1.6.2</version>
-      </extension>
-    </extensions>
     <pluginManagement>
       <plugins>
         <!-- mvn -N io.takari:maven:wrapper generates the ./mvnw script -->

--- a/pom.xml
+++ b/pom.xml
@@ -53,11 +53,12 @@
     <main.basedir>${project.basedir}</main.basedir>
 
     <armeria.version>0.84.0</armeria.version>
-    <grpc.version>1.19.0</grpc.version>
-    <protobuf.version>3.7.0</protobuf.version>
-    <unpack-proto.directory>${project.build.directory}/test/proto</unpack-proto.directory>
     <!-- This is from armeria, but be careful to avoid >= v20 apis -->
     <guava.version>27.0.1-jre</guava.version>
+
+    <!-- only used for proto interop testing -->
+    <wire.version>3.0.0-alpha01</wire.version>
+    <unpack-proto.directory>${project.build.directory}/test/proto</unpack-proto.directory>
 
     <brave.version>5.6.3</brave.version>
     <cassandra-driver-core.version>3.7.1</cassandra-driver-core.version>
@@ -359,15 +360,16 @@
           </exclusion>
         </exclusions>
       </dependency>
+
       <dependency>
         <groupId>io.zipkin.proto3</groupId>
         <artifactId>zipkin-proto3</artifactId>
         <version>0.1.0</version>
       </dependency>
       <dependency>
-        <groupId>com.google.protobuf</groupId>
-        <artifactId>protobuf-java</artifactId>
-        <version>${protobuf.version}</version>
+        <groupId>com.squareup.wire</groupId>
+        <artifactId>wire-runtime</artifactId>
+        <version>${wire.version}</version>
       </dependency>
     </dependencies>
   </dependencyManagement>
@@ -519,11 +521,11 @@
           </executions>
         </plugin>
 
-        <!-- The below plugins compile google grpc proto stubs in the test tree -->
+        <!-- The below plugins compile protobuf stubs in the indicated source tree -->
         <plugin>
           <artifactId>maven-dependency-plugin</artifactId>
           <executions>
-            <!-- protobuf-maven-plugin cannot get proto definitions from dependencies: this will -->
+            <!-- wire-maven-plugin cannot get proto definitions from dependencies: this will -->
             <execution>
               <id>unpack-proto</id>
               <phase>generate-sources</phase>
@@ -539,16 +541,23 @@
           </executions>
         </plugin>
         <plugin>
-          <groupId>org.xolstice.maven.plugins</groupId>
-          <artifactId>protobuf-maven-plugin</artifactId>
-          <version>0.6.1</version>
-          <configuration>
-            <protoSourceRoot>${project.build.directory}/main/proto</protoSourceRoot>
-            <protoTestSourceRoot>${project.build.directory}/test/proto</protoTestSourceRoot>
-            <protocArtifact>
-              com.google.protobuf:protoc:${protobuf.version}:exe:${os.detected.classifier}
-            </protocArtifact>
-          </configuration>
+          <groupId>com.squareup.wire</groupId>
+          <artifactId>wire-maven-plugin</artifactId>
+          <version>${wire.version}</version>
+          <executions>
+            <execution>
+              <phase>generate-sources</phase>
+              <goals>
+                <goal>generate-sources</goal>
+              </goals>
+              <configuration>
+                <protoSourceDirectory>${unpack-proto.directory}</protoSourceDirectory>
+                <includes>
+                  <include>zipkin.proto3.*</include>
+                </includes>
+              </configuration>
+            </execution>
+          </executions>
         </plugin>
       </plugins>
     </pluginManagement>

